src/or/routerlist.c

```diff
@@ -2758,15 +2758,15 @@ frac_nodes_with_descriptors(const smartlist_t *sl,
       total <= 0.0) {
     int n_with_descs = 0;
     SMARTLIST_FOREACH(sl, const node_t *, node, {
-      if (node_has_descriptor(node))
+      if (node_has_any_descriptor(node))
```

Comment:
We should pass for_direct_conn into this function, and use node_has_preferred_descriptor().

Then we won't bootstrap unless we have enough actual mid and exit bandwidth, even if we have mids or exits listed as our bridges. (The guard case is unchanged for non-bridge clients.)

The bridge client case could be tricky, because:
1. compute_frac_paths_available() only checks guard-flagged nodes, not bridges
2. even if it did check bridges, they don't have bandwidths
3. even if we used a weight of 1 for each bridge, we don't require 65% of bridges to be up to bootstrap

To workaround this issue, I suggest we make f_guard = 1.0 in compute_frac_paths_available() if we are using bridges, and have at least one bridge with the preferred descriptor.

Reply:
This change might cause bootstrapping issues. Let's defer it to 0.3.4 in another ticket?

Reply:
Opened trac:25886 for this.